{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:37.834286Z",
     "start_time": "2022-09-12T01:48:37.056785Z"
    }
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function, division\n",
    "%matplotlib inline\n",
    "import matplotlib.pylab as plt\n",
    "import os, sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os, shutil, glob\n",
    "import urllib\n",
    "import wget\n",
    "import subprocess\n",
    "import logging\n",
    "from ete3 import NCBITaxa\n",
    "ncbi = NCBITaxa()\n",
    "from collections import OrderedDict, Counter\n",
    "import time\n",
    "\n",
    "logging.basicConfig(level=logging.DEBUG,\n",
    "                    format='%(asctime)s [%(name)s] %(levelname)s : %(message)s')\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "#sys.path.append('/misc/workspace/amromics/amromics/')\n",
    "#from  amromics.utils.bioseq import read_sequence_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:39.021736Z",
     "start_time": "2022-09-12T01:48:39.016610Z"
    }
   },
   "outputs": [],
   "source": [
    "ncbi_root = '/misc/workspace/data/ncbi'\n",
    "ncbi_url = 'ftp://ftp.ncbi.nlm.nih.gov/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:39.856929Z",
     "start_time": "2022-09-12T01:48:39.836954Z"
    }
   },
   "outputs": [],
   "source": [
    "def download_assembly(genome_id):\n",
    "    \"\"\"\n",
    "    Download the the genome file from patric if not there already\n",
    "    \"\"\"\n",
    "    #make sure directory exist\n",
    "    directory = os.path.join(patric_root,'genomes',genome_id)\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "    genome_file = os.path.join(directory, genome_id + '.fna')\n",
    "    if os.path.isfile(genome_file):\n",
    "        logger.info('Genome {} has been downloaded'.format(genome_id))\n",
    "        return genome_file\n",
    "    genome_url = patric_url + 'genomes/' + genome_id + '/' + genome_id + '.fna'    \n",
    "    logger.info('Downloaded {}'.format(genome_id))\n",
    "    return wget.download(genome_url, genome_file)\n",
    "\n",
    "\n",
    "def run_command(cmd, timing_log=None):\n",
    "    \"\"\"\n",
    "    Run a command line, return the returning code of the command\n",
    "    :param cmd:\n",
    "    :param timing_log:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if timing_log is not None:\n",
    "        cmd = '/usr/bin/time --append -v -o {} bash -c \"{}\"'.format(timing_log, cmd)\n",
    "    #logger.info('Running \"{}'.format(cmd))\n",
    "    ret = os.system(cmd)\n",
    "    if ret != 0:\n",
    "        logger.error('Running \"{}\" returns {}'.format(cmd, ret))\n",
    "    return ret\n",
    "\n",
    "\n",
    "def get_linages(taxon_id, linage_ranks = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']):\n",
    "    lin_ids = ncbi.get_lineage(taxon_id)\n",
    "    lin_names =  ncbi.get_taxid_translator(lin_ids)\n",
    "    lin_ranks = ncbi.get_rank(lin_ids)\n",
    "    #swap key<-> value\n",
    "    lin_ranks = {lin_ranks[key]: key for key in lin_ranks}\n",
    "    \n",
    "    ret_names = [None] * len(linage_ranks)\n",
    "    ret_ids = [None] * len(linage_ranks)\n",
    "    for ii, rank in enumerate(linage_ranks):\n",
    "        if rank in lin_ranks:\n",
    "            lin_id = lin_ranks[rank]\n",
    "            ret_ids[ii] = lin_id\n",
    "            ret_names[ii] = lin_names[lin_id]\n",
    "    return ret_ids, ret_names            \n",
    "    \n",
    "\n",
    "def download_patric_assembly_curl(genome_id):\n",
    "    \"\"\"\n",
    "    Download the the genome file from patric if not there already\n",
    "    \"\"\"\n",
    "    #make sure directory exist\n",
    "    directory = os.path.join(patric_root,'genomes',genome_id)\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "    genome_file = os.path.join(directory, genome_id + '.fna')\n",
    "    genome_file_gz = os.path.join(directory, genome_id + '.fna.gz')\n",
    "    if os.path.isfile(genome_file_gz):\n",
    "        logger.info('Previously done {}'.format(genome_id))\n",
    "        return genome_file_gz\n",
    "    elif os.path.isfile(genome_file):\n",
    "        logger.info('Partly done {}'.format(genome_id))\n",
    "        cmd = 'gzip ' + genome_file\n",
    "        if run_command(cmd) != 0:\n",
    "            return None\n",
    "        else:\n",
    "            assert os.path.isfile(genome_file_gz)\n",
    "            return genome_file_gz\n",
    "    else:\n",
    "        # logger.info('Download')\n",
    "        genome_url = patric_url + 'genomes/' + genome_id + '/' + genome_id + '.fna'    \n",
    "        time.sleep(3)\n",
    "        cmd = 'curl -R -o {} {}'.format(genome_file, genome_url)\n",
    "        if run_command(cmd) != 0:\n",
    "            return None\n",
    "        cmd = 'gzip ' + genome_file\n",
    "        if run_command(cmd) != 0:\n",
    "            return None\n",
    "        else:\n",
    "            assert os.path.isfile(genome_file_gz)\n",
    "            return genome_file_gz\n",
    "\n",
    "def download_refseq_assembly_curl(refseq_dir):\n",
    "    \"\"\"\n",
    "    Download the the genome file from refseq if not there already\n",
    "    \"\"\"\n",
    "    #make sure directory exist\n",
    "    path = refseq_dir[len(ncbi_url):]\n",
    "    genome_path = os.path.join(ncbi_root, path)\n",
    "    os.makedirs(genome_path, exist_ok=True)\n",
    "    genome_id = refseq_dir.split('/')[-1]\n",
    "    \n",
    "    genome_file = os.path.join(genome_path, genome_id + '_genomic.fna.gz')\n",
    "    annotation_file = os.path.join(genome_path, genome_id + '_genomic.gbff.gz')\n",
    "    \n",
    "    genome_url = refseq_dir + '/' + genome_id + '_genomic.fna.gz'\n",
    "    gbf_url = refseq_dir + '/' + genome_id + '_genomic.gbff.gz'\n",
    "    if not os.path.isfile(genome_file):\n",
    "        cmd = 'curl --retry 10 -R -o {} {}'.format(genome_file, genome_url)\n",
    "        time.sleep(1)\n",
    "        ret = run_command(cmd)\n",
    "        if ret != 0:\n",
    "            logger.error('Error downloading {} return {}'.format(genome_url, ret))\n",
    "            genome_file = None\n",
    "        else:\n",
    "            logger.info('Downloaded {}'.format(genome_file))\n",
    "    else:\n",
    "        pass\n",
    "        #logger.info('File {} exists!'.format(genome_file))\n",
    "    \n",
    "    if not os.path.isfile(annotation_file):\n",
    "        cmd = 'curl --retry 3 -R -o {} {}'.format(annotation_file, gbf_url)\n",
    "        time.sleep(1)\n",
    "        ret = run_command(cmd)\n",
    "        if ret != 0:\n",
    "            logger.error('Error downloading {} return {}'.format(gbf_url, ret))\n",
    "            annotation_file = None \n",
    "        else:\n",
    "            logger.info('Downloaded {}'.format(annotation_file))\n",
    "    else:\n",
    "        #logger.info('File {} exists!'.format(annotation_file))\n",
    "        pass\n",
    "    return genome_file, annotation_file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data preprocessing\n",
    "\n",
    "## Get lineage information\n",
    "\n",
    "As the lineage information from patric might have been outdated, we refer to NCBI to update the lineage information. As in turns out, there are changes to NCBI taxatomy incuding taxon IDs and taxon name. In fact, we\n",
    "infor lieage information directly from taxon ID.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:41.197573Z",
     "start_time": "2022-09-12T01:48:41.192399Z"
    }
   },
   "outputs": [],
   "source": [
    "# #gl_df = pd.read_table(os.path.join(patric_root, 'RELEASE_NOTES', 'genome_lineage'),  dtype={'genome_id': str})\n",
    "\n",
    "\n",
    "# gm_df = pd.read_table(os.path.join(patric_root, 'RELEASE_NOTES', 'genome_metadata'), dtype={'genome_id': str})\n",
    "# #note that i decided to get the linage information instead of taking from gl\n",
    "# linages = gm_df['taxon_id'].apply(lambda x:get_linages(x))\n",
    "# linage_ids, linage_names  = zip(*linages)\n",
    "# superkingdom_names, phylum_names, class_names, order_names, family_names, genus_names, species_names = zip(*linage_names)\n",
    "# superkingdom_ids, phylum_ids, class_ids, order_ids, family_ids, genus_ids, species_ids = zip(*linage_ids)\n",
    "\n",
    "# gm_df['species_name'] = species_names\n",
    "# gm_df['genus_name'] = genus_names\n",
    "# gm_df['species_id'] = species_ids\n",
    "# gm_df['genus_id'] = genus_ids\n",
    "\n",
    "# #gm_df['family'] = family_names\n",
    "# #gm_df['order'] = order_names\n",
    "# #gm_df['phylum'] = phylum_names\n",
    "# #gm_df['domain'] = superkingdom_names\n",
    "\n",
    "# #(gm_df['domain'] == 'Bacteria')\n",
    "# #gm_df = gm_df[(gm_df['genome_status'] != 'Plasmid') & (gm_df['contigs'] < 500) ]\n",
    "# print(len(gm_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We grouped the data in species, and pick those relavant species. Note that many Salmonella taxa do not have a\n",
    "species name, and hence we get those from genus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:42.913580Z",
     "start_time": "2022-09-12T01:48:42.906730Z"
    }
   },
   "outputs": [],
   "source": [
    "#Look at the list of most common specices, and found:\n",
    "# Counter(gm_df['species']).most_common()\n",
    "amr_species = [    \n",
    "    # 'Mycobacterium tuberculosis', # 27993),\n",
    "    'Escherichia coli', # 23309),\n",
    "    'Streptococcus pneumoniae', # 20697),\n",
    "    'Salmonella enterica', # 18169),\n",
    "    'Staphylococcus aureus', # 12878),\n",
    "    'Klebsiella pneumoniae', # 12246),\n",
    "    'Acinetobacter baumannii', # 6936),\n",
    "    # None, # 5303),\n",
    "    'Pseudomonas aeruginosa', # 5263),\n",
    "    'Neisseria gonorrhoeae', # 5096),\n",
    "    # 'Listeria monocytogenes', # 4111),\n",
    "    # 'Campylobacter jejuni', # 2973),\n",
    "    # 'uncultured Pelagibacteraceae bacterium', # 2390),\n",
    "    # 'Clostridioides difficile', # 2345),\n",
    "     'Enterococcus faecium', # 2189),\n",
    "    # 'Streptococcus pyogenes', # 2063),\n",
    "    # 'Neisseria meningitidis', # 2001),\n",
    "    # 'Mycobacteroides abscessus', # 1734),\n",
    "    # 'Campylobacter coli', # 1724),\n",
    "    # 'Helicobacter pylori', # 1644),\n",
    "    # 'Burkholderia pseudomallei', # 1615),\n",
    "    'Enterococcus faecalis', # 1564),\n",
    "    # 'Pseudomonas viridiflava', # 1540),\n",
    "    'Shigella sonnei', # 1532),\n",
    "    # 'Vibrio parahaemolyticus', # 1474),\n",
    "    # 'Streptococcus agalactiae', # 1355),\n",
    "    'Vibrio cholerae', # 1350),\n",
    "    'Streptococcus suis', # 1319),\n",
    "    # 'Bacillus cereus', # 1252),\n",
    "    'Enterobacter cloacae', # 1187)\n",
    "    'Corynebacterium diphtheriae',\n",
    "]\n",
    "#gm_df = gm_df[(gm_df['genus_name'] == 'Salmonella') | (gm_df['species_name'].isin(amr_species))]\n",
    "\n",
    "\n",
    "amr_species = [    \n",
    "    # 'Mycobacterium tuberculosis', # 27993),\n",
    "   'Escherichia coli', # 23309),\n",
    "#=#    'Streptococcus pneumoniae', # 20697),\n",
    "#=#     'Salmonella enterica', # 18169),\n",
    "    'Staphylococcus aureus', # 12878),\n",
    "    'Klebsiella pneumoniae', # 12246),\n",
    "#=#     'Acinetobacter baumannii', # 6936),\n",
    "    # None, # 5303),\n",
    "#=#     'Pseudomonas aeruginosa', # 5263),\n",
    "#=#     'Neisseria gonorrhoeae', # 5096),\n",
    "    # 'Listeria monocytogenes', # 4111),\n",
    "    # 'Campylobacter jejuni', # 2973),\n",
    "    # 'uncultured Pelagibacteraceae bacterium', # 2390),\n",
    "    # 'Clostridioides difficile', # 2345),\n",
    "    # 'Enterococcus faecium', # 2189),\n",
    "    # 'Streptococcus pyogenes', # 2063),\n",
    "    # 'Neisseria meningitidis', # 2001),\n",
    "    # 'Mycobacteroides abscessus', # 1734),\n",
    "    # 'Campylobacter coli', # 1724),\n",
    "    # 'Helicobacter pylori', # 1644),\n",
    "    # 'Burkholderia pseudomallei', # 1615),\n",
    "#=#     'Enterococcus faecalis', # 1564),\n",
    "    # 'Pseudomonas viridiflava', # 1540),\n",
    "#=#     'Shigella sonnei', # 1532),\n",
    "    # 'Vibrio parahaemolyticus', # 1474),\n",
    "    # 'Streptococcus agalactiae', # 1355),\n",
    "    # 'Vibrio cholerae', # 1350),\n",
    "#=#     'Streptococcus suis', # 1319),\n",
    "    # 'Bacillus cereus', # 1252),\n",
    "    # 'Enterobacter cloacae', # 1187)\n",
    "###    'Corynebacterium diphtheriae',\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:50.148062Z",
     "start_time": "2022-09-12T01:48:43.849660Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: DtypeWarning: Columns (20) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2884447 was translated into 2792859\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1900 was translated into 1961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2501474 was translated into 2500546\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2758574 was translated into 2758440\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2758577 was translated into 2758573\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2841033 was translated into 2841032\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2759659 was translated into 2760083\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2891943 was translated into 2815579\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2498857 was translated into 2498858\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2498859 was translated into 2498858\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2893020 was translated into 2755382\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2503034 was translated into 2505976\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2816245 was translated into 2816244\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2816246 was translated into 2816244\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2816247 was translated into 2816244\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2898420 was translated into 2713171\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2899824 was translated into 2899822\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2899825 was translated into 2899822\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2769496 was translated into 2842200\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2902425 was translated into 2901869\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2817474 was translated into 2817475\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2904527 was translated into 82989\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2905875 was translated into 2905874\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2905876 was translated into 2905874\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2905878 was translated into 2905877\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2906071 was translated into 2765360\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2775032 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2775033 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2715211 was translated into 2758038\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 425933 was translated into 79879\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2768163 was translated into 2768161\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2768164 was translated into 2768162\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 166934 was translated into 77608\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1217700 was translated into 70347\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2792482 was translated into 2792481\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2847993 was translated into 2847991\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 307251 was translated into 298\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2902426 was translated into 2901869\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2545414 was translated into 2562451\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2810353 was translated into 2810351\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 189515 was translated into 134386\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2812557 was translated into 2812555\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2812630 was translated into 2758573\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2812897 was translated into 2726117\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2682802 was translated into 2654874\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 716759 was translated into 2944916\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2814245 was translated into 2814244\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2816858 was translated into 2816859\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2687303 was translated into 1255\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2820662 was translated into 2834768\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 68178 was translated into 1894\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 68210 was translated into 47760\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2692266 was translated into 2692265\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2696286 was translated into 2696285\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2828362 was translated into 2828363\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2828364 was translated into 2828363\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854762 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854763 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854764 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 67349 was translated into 1894\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2831676 was translated into 2831523\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2803866 was translated into 2828363\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2834769 was translated into 2834768\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2834770 was translated into 2834768\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2704091 was translated into 2559928\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2704092 was translated into 2559928\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2049333 was translated into 2917965\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 215842 was translated into 2883100\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1264596 was translated into 2878388\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2575373 was translated into 2575440\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1789652 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 741091 was translated into 2703885\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 482825 was translated into 172827\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2583375 was translated into 77608\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2584181 was translated into 2486008\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2585212 was translated into 2585211\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2847990 was translated into 2847989\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2848007 was translated into 754515\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2848008 was translated into 754515\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2586910 was translated into 1857024\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2853241 was translated into 2799571\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854754 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854756 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854757 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854758 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854759 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854760 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2854761 was translated into 2854755\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2593677 was translated into 2595054\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2725561 was translated into 2725562\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1546149 was translated into 164759\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 366526 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2596922 was translated into 2703794\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2596923 was translated into 129985\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2597824 was translated into 2597660\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 501011 was translated into 501010\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2734910 was translated into 2734909\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2734911 was translated into 2734909\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2604327 was translated into 2919747\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2735525 was translated into 2735528\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2867235 was translated into 2867236\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2867517 was translated into 2867516\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2867519 was translated into 2867518\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2867960 was translated into 2867412\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2738412 was translated into 2735723\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2738411 was translated into 2735723\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1955273 was translated into 2912308\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2742999 was translated into 2831468\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875765 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875766 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875767 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875768 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875769 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2875770 was translated into 2875764\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 516701 was translated into 141693\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2746960 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2746958 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2746962 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2746959 was translated into 2746961\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 1962644 was translated into 1945581\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2881053 was translated into 2881052\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2881054 was translated into 2881055\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2883171 was translated into 2883169\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2883172 was translated into 2883169\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2883173 was translated into 2883169\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2883174 was translated into 2883169\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/misc/usr/sw/conda/Miniconda3-py39_4.12.0/envs/amromics_dev/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:243: UserWarning: taxid 2883170 was translated into 2883169\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'Escherichia coli': 26392,\n",
       "         'Staphylococcus aureus': 13556,\n",
       "         'Klebsiella pneumoniae': 12123})"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_assemly = os.path.join(ncbi_root, 'assembly_summary_refseq.txt')\n",
    "\n",
    "\n",
    "refseq_df = pd.read_csv(refseq_assemly, sep='\\t', header=1)\n",
    "refseq_df.rename({'# assembly_accession':'assembly_accession'},axis=1, inplace=True)\n",
    "\n",
    "all_species_ids = set(refseq_df['species_taxid'])\n",
    "species_map = {}\n",
    "for taxon_id in all_species_ids:\n",
    "    try:\n",
    "        r = get_linages(taxon_id)\n",
    "        species = r[1][6]\n",
    "        genus = r[1][5]\n",
    "        domain = r[1][0]        \n",
    "        species_map[taxon_id] = (species, genus, domain)\n",
    "    except:\n",
    "        species_map[taxon_id] = (None, None, None)  \n",
    "        \n",
    "\n",
    "refseq_df['species_name'] = refseq_df['species_taxid'].apply(lambda x:species_map[x][0])\n",
    "refseq_df['genus_name']   = refseq_df['species_taxid'].apply(lambda x:species_map[x][1])\n",
    "refseq_df['domain_name']  = refseq_df['species_taxid'].apply(lambda x:species_map[x][2])\n",
    "\n",
    "#refseq_df = refseq_df[(refseq_df['genus_name'] == 'Salmonella') | (refseq_df['species_name'].isin(amr_species))].reset_index()\n",
    "df = refseq_df[(refseq_df['species_name'].isin(amr_species))].reset_index()\n",
    "\n",
    "Counter(df['species_name'])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-12T01:48:50.203279Z",
     "start_time": "2022-09-12T01:48:50.150244Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>...</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "      <th>species_name</th>\n",
       "      <th>genus_name</th>\n",
       "      <th>domain_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>1996</td>\n",
       "      <td>GCF_000166595.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02436016</td>\n",
       "      <td>AELD00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>913088</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli TW11681</td>\n",
       "      <td>strain=TW11681</td>\n",
       "      <td>...</td>\n",
       "      <td>Institute for Genome Sciences</td>\n",
       "      <td>GCA_000166595.2</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156</th>\n",
       "      <td>14702</td>\n",
       "      <td>GCF_000513035.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02304041</td>\n",
       "      <td>AVCD00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1343836</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli O157:H7 str. F8092B</td>\n",
       "      <td>strain=F8092B</td>\n",
       "      <td>...</td>\n",
       "      <td>USDA</td>\n",
       "      <td>GCA_000513035.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3989</th>\n",
       "      <td>17239</td>\n",
       "      <td>GCF_000567725.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02138579</td>\n",
       "      <td>JCNV00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1328373</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae MGH 39</td>\n",
       "      <td>strain=MGH 39</td>\n",
       "      <td>...</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>GCA_000567725.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6651</th>\n",
       "      <td>24704</td>\n",
       "      <td>GCF_000735285.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02768808</td>\n",
       "      <td>JPIQ00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=IA565</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Michigan</td>\n",
       "      <td>GCA_000735285.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6744</th>\n",
       "      <td>25961</td>\n",
       "      <td>GCF_000770035.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02951658</td>\n",
       "      <td>JPUT00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=2011EL-1370-2</td>\n",
       "      <td>...</td>\n",
       "      <td>Centers for Disease Control and Prevention</td>\n",
       "      <td>GCA_000770035.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9738</th>\n",
       "      <td>54076</td>\n",
       "      <td>GCF_001462715.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04158281</td>\n",
       "      <td>LNGY00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=K71-77</td>\n",
       "      <td>...</td>\n",
       "      <td>Norwegian National Advisory Unit on Detection ...</td>\n",
       "      <td>GCA_001462715.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9739</th>\n",
       "      <td>54077</td>\n",
       "      <td>GCF_001462735.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04158282</td>\n",
       "      <td>LNGZ00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=K66-45</td>\n",
       "      <td>...</td>\n",
       "      <td>Norwegian National Advisory Unit on Detection ...</td>\n",
       "      <td>GCA_001462735.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9748</th>\n",
       "      <td>54086</td>\n",
       "      <td>GCF_001462925.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04158294</td>\n",
       "      <td>LNHL00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=50579417</td>\n",
       "      <td>...</td>\n",
       "      <td>Norwegian National Advisory Unit on Detection ...</td>\n",
       "      <td>GCA_001462925.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10949</th>\n",
       "      <td>64627</td>\n",
       "      <td>GCF_001728785.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN04191551</td>\n",
       "      <td>LYVB00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=FRIK2069</td>\n",
       "      <td>...</td>\n",
       "      <td>US Food and Drug Administration</td>\n",
       "      <td>GCA_001728785.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19094</th>\n",
       "      <td>101926</td>\n",
       "      <td>GCF_003008295.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN08637771</td>\n",
       "      <td>PVPK00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=BA22372</td>\n",
       "      <td>...</td>\n",
       "      <td>Christian Medical College</td>\n",
       "      <td>GCA_003008295.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23555</th>\n",
       "      <td>123283</td>\n",
       "      <td>GCF_003980305.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN08161240</td>\n",
       "      <td>PNRR00000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=TW10590</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Maryland, School of Medicine</td>\n",
       "      <td>GCA_003980305.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36722</th>\n",
       "      <td>186107</td>\n",
       "      <td>GCF_016785485.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387362</td>\n",
       "      <td>JADBEX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=BS74R-D</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785485.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36723</th>\n",
       "      <td>186108</td>\n",
       "      <td>GCF_016785505.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387357</td>\n",
       "      <td>JADBFC000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=BS15R-A</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785505.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36724</th>\n",
       "      <td>186109</td>\n",
       "      <td>GCF_016785525.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387510</td>\n",
       "      <td>JADCMM000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EM20-18-50</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785525.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36725</th>\n",
       "      <td>186110</td>\n",
       "      <td>GCF_016785585.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387507</td>\n",
       "      <td>JADBFF000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EM10-18-28</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785585.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36727</th>\n",
       "      <td>186112</td>\n",
       "      <td>GCF_016785625.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387502</td>\n",
       "      <td>JADBFJ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EM06-18-14</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785625.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36731</th>\n",
       "      <td>186116</td>\n",
       "      <td>GCF_016785705.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387483</td>\n",
       "      <td>JADBFO000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EF7-18-51</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016785705.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36736</th>\n",
       "      <td>186130</td>\n",
       "      <td>GCF_016786025.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387358</td>\n",
       "      <td>JADBFB000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=BS90R-A</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786025.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36737</th>\n",
       "      <td>186131</td>\n",
       "      <td>GCF_016786035.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387495</td>\n",
       "      <td>JADBFV000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EH09-18-41</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786035.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36739</th>\n",
       "      <td>186133</td>\n",
       "      <td>GCF_016786075.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387482</td>\n",
       "      <td>JADBFP000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EF5-18-41</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786075.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36740</th>\n",
       "      <td>186134</td>\n",
       "      <td>GCF_016786085.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387487</td>\n",
       "      <td>JADCMS000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EF7-18-58</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786085.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36741</th>\n",
       "      <td>186135</td>\n",
       "      <td>GCF_016786125.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387489</td>\n",
       "      <td>JADBFY000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EH01-18-02</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786125.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36742</th>\n",
       "      <td>186136</td>\n",
       "      <td>GCF_016786145.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387494</td>\n",
       "      <td>JADBFW000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EH08-18-36</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786145.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36743</th>\n",
       "      <td>186137</td>\n",
       "      <td>GCF_016786155.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387490</td>\n",
       "      <td>JADCMY000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EH01-18-04-A</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016786155.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36752</th>\n",
       "      <td>186167</td>\n",
       "      <td>GCF_016791825.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387501</td>\n",
       "      <td>JADBFK000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EM03-18-08</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016791825.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36763</th>\n",
       "      <td>186178</td>\n",
       "      <td>GCF_016792035.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN16387496</td>\n",
       "      <td>JADBFU000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=EH10-18-47</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_016792035.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43012</th>\n",
       "      <td>213950</td>\n",
       "      <td>GCF_021136355.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016516</td>\n",
       "      <td>JAIUAO000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=BD-41</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136355.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43021</th>\n",
       "      <td>213959</td>\n",
       "      <td>GCF_021136665.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016508</td>\n",
       "      <td>JAIUAY000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=BD-13</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136665.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43022</th>\n",
       "      <td>213960</td>\n",
       "      <td>GCF_021136695.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016509</td>\n",
       "      <td>JAIUAX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=BD-14</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136695.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43024</th>\n",
       "      <td>213962</td>\n",
       "      <td>GCF_021136735.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016532</td>\n",
       "      <td>JAIUBD000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=7008.20</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136735.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43026</th>\n",
       "      <td>213964</td>\n",
       "      <td>GCF_021136775.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016533</td>\n",
       "      <td>JAIUBC000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=7011.62</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136775.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43028</th>\n",
       "      <td>213966</td>\n",
       "      <td>GCF_021136815.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016531</td>\n",
       "      <td>JAIUBE000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6911.28</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136815.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43029</th>\n",
       "      <td>213967</td>\n",
       "      <td>GCF_021136835.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016528</td>\n",
       "      <td>JAIUBH000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6711.43</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136835.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43030</th>\n",
       "      <td>213968</td>\n",
       "      <td>GCF_021136845.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016529</td>\n",
       "      <td>JAIUBG000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6712.08</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136845.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43031</th>\n",
       "      <td>213969</td>\n",
       "      <td>GCF_021136875.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016527</td>\n",
       "      <td>JAIUBI000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6710.71</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136875.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43032</th>\n",
       "      <td>213970</td>\n",
       "      <td>GCF_021136895.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016534</td>\n",
       "      <td>JAIUBJ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6709.15-I</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136895.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43035</th>\n",
       "      <td>213973</td>\n",
       "      <td>GCF_021136945.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016525</td>\n",
       "      <td>JAIUBL000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6611.48</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136945.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43036</th>\n",
       "      <td>213974</td>\n",
       "      <td>GCF_021136955.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016524</td>\n",
       "      <td>JAIUBM000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=6604.68</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021136955.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43039</th>\n",
       "      <td>213977</td>\n",
       "      <td>GCF_021137035.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016539</td>\n",
       "      <td>JAIUBS000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=5208.51</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137035.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43044</th>\n",
       "      <td>213982</td>\n",
       "      <td>GCF_021137135.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016540</td>\n",
       "      <td>JAIUBZ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=0213293972</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137135.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43046</th>\n",
       "      <td>213984</td>\n",
       "      <td>GCF_021137155.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016517</td>\n",
       "      <td>JAIUBY000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=400195-1-18</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137155.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43047</th>\n",
       "      <td>213985</td>\n",
       "      <td>GCF_021137165.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016535</td>\n",
       "      <td>JAIUBW000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=4906.28</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137165.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43048</th>\n",
       "      <td>213986</td>\n",
       "      <td>GCF_021137215.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016518</td>\n",
       "      <td>JAIUBX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=400195-2-18</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137215.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43049</th>\n",
       "      <td>213987</td>\n",
       "      <td>GCF_021137235.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016544</td>\n",
       "      <td>JAIUCB000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=1613163991</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137235.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43050</th>\n",
       "      <td>213988</td>\n",
       "      <td>GCF_021137255.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016545</td>\n",
       "      <td>JAIUCE000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=0113481141</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137255.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43051</th>\n",
       "      <td>213989</td>\n",
       "      <td>GCF_021137275.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016543</td>\n",
       "      <td>JAIUCA000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=1613258639</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137275.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43052</th>\n",
       "      <td>213990</td>\n",
       "      <td>GCF_021137295.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016542</td>\n",
       "      <td>JAIUCC000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=1613015451</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137295.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43053</th>\n",
       "      <td>213991</td>\n",
       "      <td>GCF_021137315.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN21016479</td>\n",
       "      <td>JAIUCD000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=1230_0791_KP</td>\n",
       "      <td>...</td>\n",
       "      <td>University of Bern</td>\n",
       "      <td>GCA_021137315.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48 rows  27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index assembly_accession   bioproject     biosample  \\\n",
       "121      1996    GCF_000166595.1  PRJNA224116  SAMN02436016   \n",
       "2156    14702    GCF_000513035.1  PRJNA224116  SAMN02304041   \n",
       "3989    17239    GCF_000567725.1  PRJNA224116  SAMN02138579   \n",
       "6651    24704    GCF_000735285.1  PRJNA224116  SAMN02768808   \n",
       "6744    25961    GCF_000770035.1  PRJNA224116  SAMN02951658   \n",
       "9738    54076    GCF_001462715.1  PRJNA224116  SAMN04158281   \n",
       "9739    54077    GCF_001462735.1  PRJNA224116  SAMN04158282   \n",
       "9748    54086    GCF_001462925.1  PRJNA224116  SAMN04158294   \n",
       "10949   64627    GCF_001728785.1  PRJNA224116  SAMN04191551   \n",
       "19094  101926    GCF_003008295.1  PRJNA224116  SAMN08637771   \n",
       "23555  123283    GCF_003980305.1  PRJNA224116  SAMN08161240   \n",
       "36722  186107    GCF_016785485.1  PRJNA224116  SAMN16387362   \n",
       "36723  186108    GCF_016785505.1  PRJNA224116  SAMN16387357   \n",
       "36724  186109    GCF_016785525.1  PRJNA224116  SAMN16387510   \n",
       "36725  186110    GCF_016785585.1  PRJNA224116  SAMN16387507   \n",
       "36727  186112    GCF_016785625.1  PRJNA224116  SAMN16387502   \n",
       "36731  186116    GCF_016785705.1  PRJNA224116  SAMN16387483   \n",
       "36736  186130    GCF_016786025.1  PRJNA224116  SAMN16387358   \n",
       "36737  186131    GCF_016786035.1  PRJNA224116  SAMN16387495   \n",
       "36739  186133    GCF_016786075.1  PRJNA224116  SAMN16387482   \n",
       "36740  186134    GCF_016786085.1  PRJNA224116  SAMN16387487   \n",
       "36741  186135    GCF_016786125.1  PRJNA224116  SAMN16387489   \n",
       "36742  186136    GCF_016786145.1  PRJNA224116  SAMN16387494   \n",
       "36743  186137    GCF_016786155.1  PRJNA224116  SAMN16387490   \n",
       "36752  186167    GCF_016791825.1  PRJNA224116  SAMN16387501   \n",
       "36763  186178    GCF_016792035.1  PRJNA224116  SAMN16387496   \n",
       "43012  213950    GCF_021136355.1  PRJNA224116  SAMN21016516   \n",
       "43021  213959    GCF_021136665.1  PRJNA224116  SAMN21016508   \n",
       "43022  213960    GCF_021136695.1  PRJNA224116  SAMN21016509   \n",
       "43024  213962    GCF_021136735.1  PRJNA224116  SAMN21016532   \n",
       "43026  213964    GCF_021136775.1  PRJNA224116  SAMN21016533   \n",
       "43028  213966    GCF_021136815.1  PRJNA224116  SAMN21016531   \n",
       "43029  213967    GCF_021136835.1  PRJNA224116  SAMN21016528   \n",
       "43030  213968    GCF_021136845.1  PRJNA224116  SAMN21016529   \n",
       "43031  213969    GCF_021136875.1  PRJNA224116  SAMN21016527   \n",
       "43032  213970    GCF_021136895.1  PRJNA224116  SAMN21016534   \n",
       "43035  213973    GCF_021136945.1  PRJNA224116  SAMN21016525   \n",
       "43036  213974    GCF_021136955.1  PRJNA224116  SAMN21016524   \n",
       "43039  213977    GCF_021137035.1  PRJNA224116  SAMN21016539   \n",
       "43044  213982    GCF_021137135.1  PRJNA224116  SAMN21016540   \n",
       "43046  213984    GCF_021137155.1  PRJNA224116  SAMN21016517   \n",
       "43047  213985    GCF_021137165.1  PRJNA224116  SAMN21016535   \n",
       "43048  213986    GCF_021137215.1  PRJNA224116  SAMN21016518   \n",
       "43049  213987    GCF_021137235.1  PRJNA224116  SAMN21016544   \n",
       "43050  213988    GCF_021137255.1  PRJNA224116  SAMN21016545   \n",
       "43051  213989    GCF_021137275.1  PRJNA224116  SAMN21016543   \n",
       "43052  213990    GCF_021137295.1  PRJNA224116  SAMN21016542   \n",
       "43053  213991    GCF_021137315.1  PRJNA224116  SAMN21016479   \n",
       "\n",
       "              wgs_master refseq_category    taxid  species_taxid  \\\n",
       "121       AELD00000000.1              na   913088            562   \n",
       "2156      AVCD00000000.1              na  1343836            562   \n",
       "3989      JCNV00000000.1              na  1328373            573   \n",
       "6651      JPIQ00000000.1              na      573            573   \n",
       "6744      JPUT00000000.1              na      562            562   \n",
       "9738      LNGY00000000.1              na      562            562   \n",
       "9739      LNGZ00000000.1              na      573            573   \n",
       "9748      LNHL00000000.1              na      562            562   \n",
       "10949     LYVB00000000.1              na      562            562   \n",
       "19094     PVPK00000000.1              na      562            562   \n",
       "23555     PNRR00000000.1              na      562            562   \n",
       "36722  JADBEX000000000.1              na      562            562   \n",
       "36723  JADBFC000000000.1              na      562            562   \n",
       "36724  JADCMM000000000.1              na      562            562   \n",
       "36725  JADBFF000000000.1              na      562            562   \n",
       "36727  JADBFJ000000000.1              na      562            562   \n",
       "36731  JADBFO000000000.1              na      562            562   \n",
       "36736  JADBFB000000000.1              na      562            562   \n",
       "36737  JADBFV000000000.1              na      562            562   \n",
       "36739  JADBFP000000000.1              na      562            562   \n",
       "36740  JADCMS000000000.1              na      562            562   \n",
       "36741  JADBFY000000000.1              na      562            562   \n",
       "36742  JADBFW000000000.1              na      562            562   \n",
       "36743  JADCMY000000000.1              na      562            562   \n",
       "36752  JADBFK000000000.1              na      562            562   \n",
       "36763  JADBFU000000000.1              na      562            562   \n",
       "43012  JAIUAO000000000.1              na      573            573   \n",
       "43021  JAIUAY000000000.1              na      573            573   \n",
       "43022  JAIUAX000000000.1              na      573            573   \n",
       "43024  JAIUBD000000000.1              na      573            573   \n",
       "43026  JAIUBC000000000.1              na      573            573   \n",
       "43028  JAIUBE000000000.1              na      573            573   \n",
       "43029  JAIUBH000000000.1              na      573            573   \n",
       "43030  JAIUBG000000000.1              na      573            573   \n",
       "43031  JAIUBI000000000.1              na      573            573   \n",
       "43032  JAIUBJ000000000.1              na      573            573   \n",
       "43035  JAIUBL000000000.1              na      573            573   \n",
       "43036  JAIUBM000000000.1              na      573            573   \n",
       "43039  JAIUBS000000000.1              na      573            573   \n",
       "43044  JAIUBZ000000000.1              na      573            573   \n",
       "43046  JAIUBY000000000.1              na      573            573   \n",
       "43047  JAIUBW000000000.1              na      573            573   \n",
       "43048  JAIUBX000000000.1              na      573            573   \n",
       "43049  JAIUCB000000000.1              na      573            573   \n",
       "43050  JAIUCE000000000.1              na      573            573   \n",
       "43051  JAIUCA000000000.1              na      573            573   \n",
       "43052  JAIUCC000000000.1              na      573            573   \n",
       "43053  JAIUCD000000000.1              na      573            573   \n",
       "\n",
       "                              organism_name    infraspecific_name  ...  \\\n",
       "121                Escherichia coli TW11681        strain=TW11681  ...   \n",
       "2156   Escherichia coli O157:H7 str. F8092B         strain=F8092B  ...   \n",
       "3989           Klebsiella pneumoniae MGH 39         strain=MGH 39  ...   \n",
       "6651                  Klebsiella pneumoniae          strain=IA565  ...   \n",
       "6744                       Escherichia coli  strain=2011EL-1370-2  ...   \n",
       "9738                       Escherichia coli         strain=K71-77  ...   \n",
       "9739                  Klebsiella pneumoniae         strain=K66-45  ...   \n",
       "9748                       Escherichia coli       strain=50579417  ...   \n",
       "10949                      Escherichia coli       strain=FRIK2069  ...   \n",
       "19094                      Escherichia coli        strain=BA22372  ...   \n",
       "23555                      Escherichia coli        strain=TW10590  ...   \n",
       "36722                      Escherichia coli        strain=BS74R-D  ...   \n",
       "36723                      Escherichia coli        strain=BS15R-A  ...   \n",
       "36724                      Escherichia coli     strain=EM20-18-50  ...   \n",
       "36725                      Escherichia coli     strain=EM10-18-28  ...   \n",
       "36727                      Escherichia coli     strain=EM06-18-14  ...   \n",
       "36731                      Escherichia coli      strain=EF7-18-51  ...   \n",
       "36736                      Escherichia coli        strain=BS90R-A  ...   \n",
       "36737                      Escherichia coli     strain=EH09-18-41  ...   \n",
       "36739                      Escherichia coli      strain=EF5-18-41  ...   \n",
       "36740                      Escherichia coli      strain=EF7-18-58  ...   \n",
       "36741                      Escherichia coli     strain=EH01-18-02  ...   \n",
       "36742                      Escherichia coli     strain=EH08-18-36  ...   \n",
       "36743                      Escherichia coli   strain=EH01-18-04-A  ...   \n",
       "36752                      Escherichia coli     strain=EM03-18-08  ...   \n",
       "36763                      Escherichia coli     strain=EH10-18-47  ...   \n",
       "43012                 Klebsiella pneumoniae          strain=BD-41  ...   \n",
       "43021                 Klebsiella pneumoniae          strain=BD-13  ...   \n",
       "43022                 Klebsiella pneumoniae          strain=BD-14  ...   \n",
       "43024                 Klebsiella pneumoniae        strain=7008.20  ...   \n",
       "43026                 Klebsiella pneumoniae        strain=7011.62  ...   \n",
       "43028                 Klebsiella pneumoniae        strain=6911.28  ...   \n",
       "43029                 Klebsiella pneumoniae        strain=6711.43  ...   \n",
       "43030                 Klebsiella pneumoniae        strain=6712.08  ...   \n",
       "43031                 Klebsiella pneumoniae        strain=6710.71  ...   \n",
       "43032                 Klebsiella pneumoniae      strain=6709.15-I  ...   \n",
       "43035                 Klebsiella pneumoniae        strain=6611.48  ...   \n",
       "43036                 Klebsiella pneumoniae        strain=6604.68  ...   \n",
       "43039                 Klebsiella pneumoniae        strain=5208.51  ...   \n",
       "43044                 Klebsiella pneumoniae     strain=0213293972  ...   \n",
       "43046                 Klebsiella pneumoniae    strain=400195-1-18  ...   \n",
       "43047                 Klebsiella pneumoniae        strain=4906.28  ...   \n",
       "43048                 Klebsiella pneumoniae    strain=400195-2-18  ...   \n",
       "43049                 Klebsiella pneumoniae     strain=1613163991  ...   \n",
       "43050                 Klebsiella pneumoniae     strain=0113481141  ...   \n",
       "43051                 Klebsiella pneumoniae     strain=1613258639  ...   \n",
       "43052                 Klebsiella pneumoniae     strain=1613015451  ...   \n",
       "43053                 Klebsiella pneumoniae   strain=1230_0791_KP  ...   \n",
       "\n",
       "                                               submitter  gbrs_paired_asm  \\\n",
       "121                        Institute for Genome Sciences  GCA_000166595.2   \n",
       "2156                                                USDA  GCA_000513035.1   \n",
       "3989                                     Broad Institute  GCA_000567725.1   \n",
       "6651                              University of Michigan  GCA_000735285.1   \n",
       "6744          Centers for Disease Control and Prevention  GCA_000770035.1   \n",
       "9738   Norwegian National Advisory Unit on Detection ...  GCA_001462715.1   \n",
       "9739   Norwegian National Advisory Unit on Detection ...  GCA_001462735.1   \n",
       "9748   Norwegian National Advisory Unit on Detection ...  GCA_001462925.1   \n",
       "10949                    US Food and Drug Administration  GCA_001728785.1   \n",
       "19094                          Christian Medical College  GCA_003008295.1   \n",
       "23555         University of Maryland, School of Medicine  GCA_003980305.1   \n",
       "36722                                 University of Bern  GCA_016785485.1   \n",
       "36723                                 University of Bern  GCA_016785505.1   \n",
       "36724                                 University of Bern  GCA_016785525.1   \n",
       "36725                                 University of Bern  GCA_016785585.1   \n",
       "36727                                 University of Bern  GCA_016785625.1   \n",
       "36731                                 University of Bern  GCA_016785705.1   \n",
       "36736                                 University of Bern  GCA_016786025.1   \n",
       "36737                                 University of Bern  GCA_016786035.1   \n",
       "36739                                 University of Bern  GCA_016786075.1   \n",
       "36740                                 University of Bern  GCA_016786085.1   \n",
       "36741                                 University of Bern  GCA_016786125.1   \n",
       "36742                                 University of Bern  GCA_016786145.1   \n",
       "36743                                 University of Bern  GCA_016786155.1   \n",
       "36752                                 University of Bern  GCA_016791825.1   \n",
       "36763                                 University of Bern  GCA_016792035.1   \n",
       "43012                                 University of Bern  GCA_021136355.1   \n",
       "43021                                 University of Bern  GCA_021136665.1   \n",
       "43022                                 University of Bern  GCA_021136695.1   \n",
       "43024                                 University of Bern  GCA_021136735.1   \n",
       "43026                                 University of Bern  GCA_021136775.1   \n",
       "43028                                 University of Bern  GCA_021136815.1   \n",
       "43029                                 University of Bern  GCA_021136835.1   \n",
       "43030                                 University of Bern  GCA_021136845.1   \n",
       "43031                                 University of Bern  GCA_021136875.1   \n",
       "43032                                 University of Bern  GCA_021136895.1   \n",
       "43035                                 University of Bern  GCA_021136945.1   \n",
       "43036                                 University of Bern  GCA_021136955.1   \n",
       "43039                                 University of Bern  GCA_021137035.1   \n",
       "43044                                 University of Bern  GCA_021137135.1   \n",
       "43046                                 University of Bern  GCA_021137155.1   \n",
       "43047                                 University of Bern  GCA_021137165.1   \n",
       "43048                                 University of Bern  GCA_021137215.1   \n",
       "43049                                 University of Bern  GCA_021137235.1   \n",
       "43050                                 University of Bern  GCA_021137255.1   \n",
       "43051                                 University of Bern  GCA_021137275.1   \n",
       "43052                                 University of Bern  GCA_021137295.1   \n",
       "43053                                 University of Bern  GCA_021137315.1   \n",
       "\n",
       "      paired_asm_comp                                           ftp_path  \\\n",
       "121         identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "2156        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "3989        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "6651        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "6744        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "9738        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "9739        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "9748        identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "10949       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "19094       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "23555       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36722       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36723       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36724       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36725       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36727       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36731       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36736       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36737       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36739       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36740       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36741       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36742       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36743       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36752       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "36763       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43012       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43021       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43022       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43024       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43026       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43028       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43029       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43030       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43031       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43032       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43035       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43036       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43039       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43044       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43046       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43047       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43048       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43049       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43050       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43051       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43052       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "43053       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...   \n",
       "\n",
       "      excluded_from_refseq relation_to_type_material asm_not_live_date  \\\n",
       "121                    NaN                       NaN                na   \n",
       "2156                   NaN                       NaN                na   \n",
       "3989                   NaN                       NaN                na   \n",
       "6651                   NaN                       NaN                na   \n",
       "6744                   NaN                       NaN                na   \n",
       "9738                   NaN                       NaN                na   \n",
       "9739                   NaN                       NaN                na   \n",
       "9748                   NaN                       NaN                na   \n",
       "10949                  NaN                       NaN                na   \n",
       "19094                  NaN                       NaN                na   \n",
       "23555                  NaN                       NaN                na   \n",
       "36722                  NaN                       NaN                na   \n",
       "36723                  NaN                       NaN                na   \n",
       "36724                  NaN                       NaN                na   \n",
       "36725                  NaN                       NaN                na   \n",
       "36727                  NaN                       NaN                na   \n",
       "36731                  NaN                       NaN                na   \n",
       "36736                  NaN                       NaN                na   \n",
       "36737                  NaN                       NaN                na   \n",
       "36739                  NaN                       NaN                na   \n",
       "36740                  NaN                       NaN                na   \n",
       "36741                  NaN                       NaN                na   \n",
       "36742                  NaN                       NaN                na   \n",
       "36743                  NaN                       NaN                na   \n",
       "36752                  NaN                       NaN                na   \n",
       "36763                  NaN                       NaN                na   \n",
       "43012                  NaN                       NaN                na   \n",
       "43021                  NaN                       NaN                na   \n",
       "43022                  NaN                       NaN                na   \n",
       "43024                  NaN                       NaN                na   \n",
       "43026                  NaN                       NaN                na   \n",
       "43028                  NaN                       NaN                na   \n",
       "43029                  NaN                       NaN                na   \n",
       "43030                  NaN                       NaN                na   \n",
       "43031                  NaN                       NaN                na   \n",
       "43032                  NaN                       NaN                na   \n",
       "43035                  NaN                       NaN                na   \n",
       "43036                  NaN                       NaN                na   \n",
       "43039                  NaN                       NaN                na   \n",
       "43044                  NaN                       NaN                na   \n",
       "43046                  NaN                       NaN                na   \n",
       "43047                  NaN                       NaN                na   \n",
       "43048                  NaN                       NaN                na   \n",
       "43049                  NaN                       NaN                na   \n",
       "43050                  NaN                       NaN                na   \n",
       "43051                  NaN                       NaN                na   \n",
       "43052                  NaN                       NaN                na   \n",
       "43053                  NaN                       NaN                na   \n",
       "\n",
       "                species_name   genus_name domain_name  \n",
       "121         Escherichia coli  Escherichia    Bacteria  \n",
       "2156        Escherichia coli  Escherichia    Bacteria  \n",
       "3989   Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "6651   Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "6744        Escherichia coli  Escherichia    Bacteria  \n",
       "9738        Escherichia coli  Escherichia    Bacteria  \n",
       "9739   Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "9748        Escherichia coli  Escherichia    Bacteria  \n",
       "10949       Escherichia coli  Escherichia    Bacteria  \n",
       "19094       Escherichia coli  Escherichia    Bacteria  \n",
       "23555       Escherichia coli  Escherichia    Bacteria  \n",
       "36722       Escherichia coli  Escherichia    Bacteria  \n",
       "36723       Escherichia coli  Escherichia    Bacteria  \n",
       "36724       Escherichia coli  Escherichia    Bacteria  \n",
       "36725       Escherichia coli  Escherichia    Bacteria  \n",
       "36727       Escherichia coli  Escherichia    Bacteria  \n",
       "36731       Escherichia coli  Escherichia    Bacteria  \n",
       "36736       Escherichia coli  Escherichia    Bacteria  \n",
       "36737       Escherichia coli  Escherichia    Bacteria  \n",
       "36739       Escherichia coli  Escherichia    Bacteria  \n",
       "36740       Escherichia coli  Escherichia    Bacteria  \n",
       "36741       Escherichia coli  Escherichia    Bacteria  \n",
       "36742       Escherichia coli  Escherichia    Bacteria  \n",
       "36743       Escherichia coli  Escherichia    Bacteria  \n",
       "36752       Escherichia coli  Escherichia    Bacteria  \n",
       "36763       Escherichia coli  Escherichia    Bacteria  \n",
       "43012  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43021  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43022  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43024  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43026  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43028  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43029  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43030  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43031  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43032  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43035  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43036  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43039  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43044  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43046  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43047  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43048  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43049  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43050  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43051  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43052  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "43053  Klebsiella pneumoniae   Klebsiella    Bacteria  \n",
       "\n",
       "[48 rows x 27 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_df = df[df.assembly_level == 'Complete Genome']\n",
    "not_complete_df = df[df.assembly_level != 'Complete Genome']\n",
    "#[['assembly_accession', 'biosample', 'infraspecific_name','seq_rel_date', 'asm_name', 'ftp_path']]#.to_csv(None, index=False)\n",
    "not_complete_df = not_complete_df[not_complete_df['biosample'].isin(complete_df['biosample'])]\n",
    "\n",
    "not_complete_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "470\n"
     ]
    }
   ],
   "source": [
    "# for i in range(1):\n",
    "#     count = 0\n",
    "#     for ii,row in complete_df.iterrows():    \n",
    "#         genome_id = row['assembly_accession'] + '_' + row['asm_name']\n",
    "#         genome_id = genome_id.replace(':','_')\n",
    "#         genome_id = genome_id.replace(' ','_')\n",
    "#         genome_id = genome_id.replace('/','_')\n",
    "#         genome_id = genome_id.replace('#','_')\n",
    "\n",
    "#         genome_file, annotation_file = download_refseq_assembly_curl(row['ftp_path'])\n",
    "#         if (genome_file is None) or (annotation_file is None):        \n",
    "#             break\n",
    "#         # print(genome_id, row['species_name'])\n",
    "#         count += 1\n",
    "#     print(count)\n",
    "#     #time.sleep(120)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#look at pyspoa notebook to see how to align a test sequence into the msa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:amromics_dev]",
   "language": "python",
   "name": "conda-env-amromics_dev-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
